File: cpp/src/bench.cpp
// Backend-facing half of the benchmark mode. The statistics, hashes and timestamp are the portable
// mac/Sources/YOLOMasterCore/bench_stats.* shared with the Swift package.
#include "bench.hpp"
#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdio>
#include <cstring>
#include <ctime>
#include <fstream>
#include <numeric>
#include <thread>
#ifndef _WIN32
#include <sys/utsname.h>
#include <unistd.h>
#endif
#ifdef __linux__
#include <sys/sysinfo.h>
#endif

namespace yolomaster::bench {

using clk = std::chrono::steady_clock;

nlohmann::json to_json(const StageStats& s) {
    return {{"n", s.n}, {"mean", s.mean}, {"median", s.median}, {"p90", s.p90}, {"p95", s.p95},
            {"p99", s.p99}, {"min", s.min}, {"max", s.max}};
}

void Samples::add(const Backend& be, size_t ndets) {
    pre.push_back(be.pre_ms); infer.push_back(be.infer_ms); post.push_back(be.post_ms);
    total.push_back(be.pre_ms + be.infer_ms + be.post_ms);
    frames++; total_dets += static_cast<long>(ndets);
}

static cv::Mat probe_image(int imgsz) {
    return cv::Mat(imgsz, imgsz, CV_8UC3, cv::Scalar(114, 114, 114));
}

// forward_raw(decode=false) is the phone Bench tab's inferOnly; backends without it (TensorRT,
// MNN) fall back to a full infer() and say so.
static bool probe_once(Backend& be, const cv::Mat& probe, const Config& cfg, bool& infer_only) {
    if (infer_only) {
        try { be.forward_raw(probe, cfg, /*decode=*/false); return true; }
        catch (const std::exception&) { infer_only = false; }
    }
    try { (void)be.infer(probe, cfg); return true; } catch (const std::exception&) { return false; }
}

ColdResult cold_run(Backend& be, const Config& cfg, int warmup, int iters) {
    ColdResult r;
    const cv::Mat probe = probe_image(cfg.imgsz);
    bool infer_only = true;
    for (int i = 0; i < warmup; ++i) if (!probe_once(be, probe, cfg, infer_only)) break;
    std::vector<double> v; v.reserve(std::max(iters, 0));
    for (int i = 0; i < iters; ++i) {
        if (!probe_once(be, probe, cfg, infer_only)) break;
        v.push_back(be.infer_ms);
    }
    r.infer_ms = reduce(v);
    r.probe_mode = infer_only ? "infer_only" : "full";
    return r;
}

SustainedResult sustained_loop(Backend& be, const Config& cfg, int warmup, double minutes,
                               int cold_iters, const std::atomic<bool>* cancel) {
    SustainedResult r;
    const cv::Mat probe = probe_image(cfg.imgsz);
    bool infer_only = true;
    for (int i = 0; i < warmup; ++i) if (!probe_once(be, probe, cfg, infer_only)) break;
    std::vector<double> all, second;
    const auto t0 = clk::now();
    auto sec_start = t0;
    const double budget_s = minutes * 60.0;
    while (true) {
        if (cancel && cancel->load()) break;
        const double elapsed = std::chrono::duration<double>(clk::now() - t0).count();
        if (elapsed >= budget_s && static_cast<int>(all.size()) >= cold_iters) break;
        if (!probe_once(be, probe, cfg, infer_only)) break;
        all.push_back(be.infer_ms); second.push_back(be.infer_ms);
        if (std::chrono::duration<double>(clk::now() - sec_start).count() >= 1.0) {
            r.sparkline.push_back(reduce(second).median);
            second.clear(); sec_start = clk::now();
        }
    }
    if (!second.empty()) r.sparkline.push_back(reduce(second).median);
    r.duration_s = std::chrono::duration<double>(clk::now() - t0).count();
    r.infer_ms = reduce(all);
    const SustainedSummary s = summarize_sustained(all, cold_iters);
    r.cold_median_ms = s.cold_median_ms; r.sustained_median_ms = s.sustained_median_ms; r.throttle_pct = s.throttle_pct;
    r.probe_mode = infer_only ? "infer_only" : "full";
    return r;
}

static std::string read_cpu_model() {
    std::ifstream f("/proc/cpuinfo");
    std::string line;
    while (std::getline(f, line)) {
        if (line.rfind("model name", 0) == 0 || line.rfind("Hardware", 0) == 0) {
            const auto c = line.find(':');
            if (c != std::string::npos) {
                std::string v = line.substr(c + 1);
                const auto b = v.find_first_not_of(" \t");
                return b == std::string::npos ? "" : v.substr(b);
            }
        }
    }
    return "";
}

static std::string popen_line(const char* cmd) {
    std::string out;
#ifndef _WIN32
    FILE* p = popen(cmd, "r");
    if (!p) return out;
    char buf[256];
    if (fgets(buf, sizeof(buf), p)) out = buf;
    pclose(p);
    while (!out.empty() && (out.back() == '\n' || out.back() == '\r')) out.pop_back();
#endif
    return out;
}

EnvInfo collect_env(const Backend& be, int threads) {
    EnvInfo e;
#ifndef _WIN32
    char h[256] = {0};
    if (gethostname(h, sizeof(h) - 1) == 0) e.host = h;
    struct utsname u{};
    if (uname(&u) == 0) e.os = std::string(u.sysname) + " " + u.release + " " + u.machine;
#endif
    e.cpu_model = read_cpu_model();
    e.cpu_count = static_cast<int>(std::thread::hardware_concurrency());
#ifdef __linux__
    struct sysinfo si{};
    if (sysinfo(&si) == 0) e.mem_bytes = static_cast<long long>(si.totalram) * si.mem_unit;
#endif
    e.gpu_name = be.device_name();
    if (e.gpu_name.empty()) {
        const std::string ep = be.active_ep;
        if (ep.find("CUDA") != std::string::npos || ep.find("cuda") != std::string::npos ||
            ep.find("TRT") != std::string::npos || ep.find("trt") != std::string::npos ||
            ep.find("TensorRT") != std::string::npos)
            e.gpu_name = popen_line("nvidia-smi --query-gpu=name --format=csv,noheader 2>/dev/null");
    }
    e.threads = threads;
#ifdef YM_GIT_COMMIT
    e.git_commit = YM_GIT_COMMIT;
#endif
#ifdef YM_BUILD_FLAGS
    e.build_flags = YM_BUILD_FLAGS;
#endif
#ifdef YM_VERSION
    e.version = YM_VERSION;
#endif
    return e;
}

ModelInfo model_info(const Backend& be, const std::string& path, const std::string& backend,
                     const std::string& precision, const Config& cfg) {
    ModelInfo m;
    std::string p = path;
    while (p.size() > 1 && (p.back() == '/' || p.back() == '\\')) p.pop_back();
    const auto slash = p.find_last_of("/\\");
    std::string base = slash == std::string::npos ? p : p.substr(slash + 1);
    const auto dot = base.find_last_of('.');
    m.id = (dot == std::string::npos || dot == 0) ? base : base.substr(0, dot);
    m.path = path; m.backend = backend; m.runtime = be.runtime_name();
    m.execution_provider = be.active_ep; m.ep_note = be.ep_note; m.precision = precision;
    m.nc = cfg.num_classes(); m.is_seg = be.is_seg(); m.imgsz = cfg.imgsz;
    return m;
}

nlohmann::json to_json(const BenchResult& r) {
    nlohmann::json j;
    j["schema_version"] = kSchema;
    j["timestamp"] = r.timestamp.empty() ? timestamp_utc() : r.timestamp;
    j["tool"] = r.tool;
    j["model"] = {{"id", r.model.id}, {"path", r.model.path}, {"backend", r.model.backend},
                  {"runtime", r.model.runtime}, {"execution_provider", r.model.execution_provider},
                  {"ep_note", r.model.ep_note}, {"precision", r.model.precision}, {"nc", r.model.nc},
                  {"is_seg", r.model.is_seg}, {"imgsz", r.model.imgsz}};
    j["environment"] = {{"host", r.env.host}, {"os", r.env.os}, {"cpu_model", r.env.cpu_model},
                        {"cpu_count", r.env.cpu_count}, {"mem_bytes", r.env.mem_bytes}, {"gpu_name", r.env.gpu_name},
                        {"threads", r.env.threads}, {"git_commit", r.env.git_commit},
                        {"build_flags", r.env.build_flags}, {"version", r.env.version}};
    j["protocol"] = {{"mode", r.protocol.mode}, {"conf", r.protocol.conf}, {"iou", r.protocol.iou},
                     {"max_det", r.protocol.max_det}, {"multi_label", r.protocol.multi_label},
                     {"slicing", r.protocol.slicing}, {"tile_size", r.protocol.tile_size},
                     {"warmup", r.protocol.warmup}, {"iters", r.protocol.iters}, {"minutes", r.protocol.minutes},
                     {"probe", r.protocol.probe}, {"probe_mode", r.protocol.probe_mode},
                     {"dataset", r.protocol.dataset}, {"image_count", r.protocol.image_count},
                     {"image_list_sha256", r.protocol.image_list_sha256}};
    j["stats_convention"] = "floor_rank";
    if (r.has_cold) j["cold"] = {{"infer_ms", to_json(r.cold.infer_ms)}, {"probe_mode", r.cold.probe_mode}};
    if (r.has_sustained)
        j["sustained"] = {{"infer_ms", to_json(r.sustained.infer_ms)}, {"cold_median_ms", r.sustained.cold_median_ms},
                          {"sustained_median_ms", r.sustained.sustained_median_ms},
                          {"throttle_pct", r.sustained.throttle_pct}, {"sparkline", r.sustained.sparkline},
                          {"duration_s", r.sustained.duration_s}, {"probe_mode", r.sustained.probe_mode}};
    if (r.has_dataset)
        j["dataset"] = {{"frames", r.dataset.frames}, {"total_dets", r.dataset.total_dets},
                        {"pre_ms", to_json(r.dataset.pre_ms)}, {"infer_ms", to_json(r.dataset.infer_ms)},
                        {"post_ms", to_json(r.dataset.post_ms)}, {"total_ms", to_json(r.dataset.total_ms)},
                        {"model_fps", r.dataset.model_fps}, {"wall_s", r.dataset.wall_s}};
    if (r.accuracy.present) {
        nlohmann::json pc = nlohmann::json::array();
        for (const auto& c : r.accuracy.map.per_class)
            pc.push_back({{"class_id", c.cls}, {"n_gt", c.n_gt}, {"n_pred", c.n_pred},
                          {"ap50", c.ap50()}, {"ap5095", c.ap5095()}});
        j["accuracy"] = {{"protocol", {{"conf", r.accuracy.conf}, {"iou", r.accuracy.iou},
                                       {"max_det", r.accuracy.max_det}, {"multi_label", r.accuracy.multi_label}}},
                         {"labels", r.accuracy.labels}, {"images", r.accuracy.map.images},
                         {"map50", r.accuracy.map.map50}, {"map5095", r.accuracy.map.map5095},
                         {"per_class", pc}, {"timings", {{"infer_ms", to_json(r.accuracy.infer_ms)}}}};
    }
    return j;
}

} // namespace yolomaster::bench

File: mac/Sources/YOLOMasterKit/Bench.swift
// Benchmark mode: the `yolomaster-bench/v1` document the Linux CLI and server emit, produced on
// macOS from the Core ML path. Same statistics (floor-rank percentiles through the portable core),
// same probe (gray 114 at the model input size), same sustained definition (median of the slowest
// quarter vs the cold median), so a Mac row sits next to a Linux or phone row in one table.
//
// Cold sweep = `Detector.inferOnly` (Core ML prediction alone, probe_mode "infer_only"). The
// dataset pass collects Result.preMs / inferMs / postMs per image; the accuracy pass runs the val
// protocol (conf 0.001, IoU 0.7, max_det 300; the Kit's decode is multi-label already) and scores it
// in process with the txt-rounded values, so the number equals scoring a --save-txt dump.
import Foundation
import CoreGraphics
import Metal

public enum BenchMode: String, CaseIterable, Sendable { case off, cold, sustained }

// MARK: - the document (mirrors cpp/src/bench.cpp to_json key for key)

public struct BenchDocument: Codable {
    public struct Model: Codable {
        public var id, path, backend, runtime, execution_provider, ep_note, precision: String
        public var nc: Int; public var is_seg: Bool; public var imgsz: Int
    }
    public struct Environment: Codable {
        public var host, os, cpu_model: String
        public var cpu_count: Int; public var mem_bytes: Int64
        public var gpu_name: String; public var threads: Int
        public var git_commit, build_flags, version: String
    }
    public struct ProtocolInfo: Codable {
        public var mode: String; public var conf, iou: Float; public var max_det: Int; public var multi_label: Bool
        public var slicing: String; public var tile_size: Int
        public var warmup, iters: Int; public var minutes: Double
        public var probe, probe_mode: String
        public var dataset: String; public var image_count: Int; public var image_list_sha256: String
        public init(mode: String, conf: Float, iou: Float, max_det: Int, multi_label: Bool, slicing: String, tile_size: Int,
                    warmup: Int, iters: Int, minutes: Double, probe: String, probe_mode: String,
                    dataset: String, image_count: Int, image_list_sha256: String) {
            self.mode = mode; self.conf = conf; self.iou = iou; self.max_det = max_det; self.multi_label = multi_label
            self.slicing = slicing; self.tile_size = tile_size; self.warmup = warmup; self.iters = iters; self.minutes = minutes
            self.probe = probe; self.probe_mode = probe_mode; self.dataset = dataset; self.image_count = image_count
            self.image_list_sha256 = image_list_sha256
        }
    }
    public struct Cold: Codable { public var infer_ms: StageStats; public var probe_mode: String }
    public struct Sustained: Codable {
        public var infer_ms: StageStats
        public var cold_median_ms, sustained_median_ms, throttle_pct: Double
        public var sparkline: [Double]; public var duration_s: Double; public var probe_mode: String
        /// macOS / iOS only: ProcessInfo thermal state samples (nominal, fair, serious, critical) once per second.
        public var thermal: [String]?
    }
    public struct Dataset: Codable {
        public var frames, total_dets: Int
        public var pre_ms, infer_ms, post_ms, total_ms: StageStats
        public var model_fps, wall_s: Double
    }
    public struct AccuracyProtocol: Codable { public var conf, iou: Float; public var max_det: Int; public var multi_label: Bool }
    public struct PerClass: Codable { public var class_id, n_gt, n_pred: Int; public var ap50, ap5095: Double }
    public struct Accuracy: Codable {
        public var `protocol`: AccuracyProtocol
        public var labels: String; public var images: Int
        public var map50, map5095: Double
        public var per_class: [PerClass]
        public var timings: [String: StageStats]
    }

    public var schema_version = "yolomaster-bench/v1"
    public var timestamp: String
    public var tool: String
    public var model: Model
    public var environment: Environment
    public var `protocol`: ProtocolInfo
    public var stats_convention = "floor_rank"
    public var cold: Cold?
    public var sustained: Sustained?
    public var dataset: Dataset?
    public var accuracy: Accuracy?

    public init(timestamp: String, tool: String, model: Model, environment: Environment, protocol proto: ProtocolInfo,
                cold: Cold? = nil, sustained: Sustained? = nil, dataset: Dataset? = nil, accuracy: Accuracy? = nil) {
        self.timestamp = timestamp; self.tool = tool; self.model = model; self.environment = environment
        self.protocol = proto; self.cold = cold; self.sustained = sustained; self.dataset = dataset; self.accuracy = accuracy
    }

    public func json() throws -> Data {
        let enc = JSONEncoder()
        enc.outputFormatting = [.prettyPrinted, .sortedKeys]
        return try enc.encode(self)
    }
}

// MARK: - environment / model cards

public enum BenchEnvironment {
    static func sysctlString(_ name: String) -> String {
        var size = 0
        guard sysctlbyname(name, nil, &size, nil, 0) == 0, size > 0 else { return "" }
        var buf = [CChar](repeating: 0, count: size)
        guard sysctlbyname(name, &buf, &size, nil, 0) == 0 else { return "" }
        return String(cString: buf)
    }
    static func sysctlInt64(_ name: String) -> Int64 {
        var v: Int64 = 0; var size = MemoryLayout<Int64>.size
        return sysctlbyname(name, &v, &size, nil, 0) == 0 ? v : 0
    }
    /// Runtime version: the app / package bundle version, else the repo VERSION file next to the
    /// package (development builds), else "".
    public static var version: String {
        if let v = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String, !v.isEmpty { return v }
        let here = URL(fileURLWithPath: #filePath)   // mac/Sources/YOLOMasterKit/Bench.swift -> repo root VERSION
        let root = here.deletingLastPathComponent().deletingLastPathComponent().deletingLastPathComponent().deletingLastPathComponent()
        if let s = try? String(contentsOf: root.appendingPathComponent("VERSION"), encoding: .utf8) {
            return s.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        return ""
    }
    public static func collect(threads: Int = 0) -> BenchDocument.Environment {
        let pi = ProcessInfo.processInfo
        let osv = pi.operatingSystemVersion
        let cpu = sysctlString("machdep.cpu.brand_string")
        let gpu = MTLCreateSystemDefaultDevice()?.name ?? ""
        return BenchDocument.Environment(
            host: pi.hostName, os: "macOS \(osv.majorVersion).\(osv.minorVersion).\(osv.patchVersion) \(sysctlString("hw.machine"))",
            cpu_model: cpu, cpu_count: pi.processorCount, mem_bytes: sysctlInt64("hw.memsize"),
            gpu_name: gpu, threads: threads, git_commit: "", build_flags: "coreml", version: version)
    }
    public static func model(_ det: Detector) -> BenchDocument.Model {
        let path = det.modelURL.path
        let id = det.modelURL.deletingPathExtension().lastPathComponent
        let ep: String = {
            switch det.computeMode {
            case .all: return "CoreML-ANE"
            case .cpuAndGPU: return "CoreML-GPU"
            case .cpu: return "CoreML-CPU"
            }
        }()
        let precision = det.metadata["precision"] ?? det.metadata["quantization"] ?? "unknown"
        return BenchDocument.Model(id: id, path: path, backend: "coreml", runtime: "coreml", execution_provider: ep,
                                   ep_note: "", precision: precision, nc: det.nc, is_seg: det.isSegment, imgsz: det.imgsz)
    }
}

// MARK: - runner

public struct BenchSamples {
    public var pre: [Double] = [], infer: [Double] = [], post: [Double] = [], total: [Double] = []
    public var frames = 0, totalDets = 0
    public init() {}
    public mutating func add(_ r: Detector.Result) {
        pre.append(r.preMs); infer.append(r.inferMs); post.append(r.postMs); total.append(r.preMs + r.inferMs + r.postMs)
        frames += 1; totalDets += r.detections.count
    }
    public var avgTotalMs: Double { frames > 0 ? total.reduce(0, +) / Double(frames) : 0 }
    public func dataset(wallS: Double) -> BenchDocument.Dataset {
        BenchDocument.Dataset(frames: frames, total_dets: totalDets, pre_ms: StageStats(pre), infer_ms: StageStats(infer),
                              post_ms: StageStats(post), total_ms: StageStats(total),
                              model_fps: avgTotalMs > 0 ? 1000 / avgTotalMs : 0, wall_s: wallS)
    }
}

public enum BenchRunner {
    /// The probe the Linux runtime uses: a gray-114 image at the model input size (letterbox is a no-op).
    public static func probeImage(_ imgsz: Int) -> CGImage? {
        var px = [UInt8](repeating: 114, count: imgsz * imgsz * 4)
        for i in stride(from: 3, to: px.count, by: 4) { px[i] = 255 }
        return px.withUnsafeMutableBytes { raw -> CGImage? in
            guard let ctx = CGContext(data: raw.baseAddress, width: imgsz, height: imgsz, bitsPerComponent: 8,
                                      bytesPerRow: imgsz * 4, space: CGColorSpaceCreateDeviceRGB(),
                                      bitmapInfo: CGImageAlphaInfo.noneSkipLast.rawValue) else { return nil }
            return ctx.makeImage()
        }
    }

    /// `warmup` untimed then `iters` timed Core ML predictions on the probe (probe_mode "infer_only").
    public static func coldSweep(_ det: Detector, warmup: Int, iters: Int) -> BenchDocument.Cold {
        guard let probe = probeImage(det.imgsz) else { return BenchDocument.Cold(infer_ms: StageStats([]), probe_mode: "infer_only") }
        for _ in 0..<max(warmup, 0) { _ = try? det.inferOnly(probe) }
        var v: [Double] = []
        v.reserveCapacity(max(iters, 0))
        for _ in 0..<max(iters, 0) {
            autoreleasepool { if let t = try? det.inferOnly(probe) { v.append(t) } }
        }
        return BenchDocument.Cold(infer_ms: StageStats(v), probe_mode: "infer_only")
    }

    /// Timed loop for `minutes` after `warmup`; cold baseline = median of the first `coldIters`;
    /// one sparkline median and one thermal-state sample per second. `cancel()` stops early.
    public static func sustainedLoop(_ det: Detector, warmup: Int, minutes: Double, coldIters: Int = 50,
                                     cancel: (() -> Bool)? = nil, tick: ((Double, Double) -> Void)? = nil) -> BenchDocument.Sustained {
        var r = BenchDocument.Sustained(infer_ms: StageStats([]), cold_median_ms: 0, sustained_median_ms: 0, throttle_pct: 0,
                                        sparkline: [], duration_s: 0, probe_mode: "infer_only", thermal: [])
        guard let probe = probeImage(det.imgsz) else { return r }
        for _ in 0..<max(warmup, 0) { _ = try? det.inferOnly(probe) }
        var all: [Double] = [], second: [Double] = []
        let t0 = Date(); var secStart = t0
        let budget = minutes * 60
        while true {
            if cancel?() == true { break }
            let elapsed = Date().timeIntervalSince(t0)
            if elapsed >= budget && all.count >= coldIters { break }
            guard let t = (try? autoreleasepool { try det.inferOnly(probe) }) else { break }
            all.append(t); second.append(t)
            if Date().timeIntervalSince(secStart) >= 1 {
                r.sparkline.append(StageStats(second).median)
                r.thermal?.append(thermalName(ProcessInfo.processInfo.thermalState))
                tick?(elapsed, StageStats(second).median)
                second.removeAll(); secStart = Date()
            }
        }
        if !second.isEmpty { r.sparkline.append(StageStats(second).median) }
        r.duration_s = Date().timeIntervalSince(t0)
        r.infer_ms = StageStats(all)
        let s = SustainedSummary(all, coldIters: coldIters)
        r.cold_median_ms = s.coldMedianMs; r.sustained_median_ms = s.sustainedMedianMs; r.throttle_pct = s.throttlePct
        return r
    }

    public static func thermalName(_ s: ProcessInfo.ThermalState) -> String {
        switch s { case .nominal: return "nominal"; case .fair: return "fair"; case .serious: return "serious"; case .critical: return "critical"
        @unknown default: return "unknown" }
    }
}

// MARK: - accuracy pass (the Linux --accuracy protocol)

public struct AccuracyProtocolValues {
    /// conf 0.001 / IoU 0.7 / max_det 300, multi-label. The Kit's decode keeps candidates with
    /// score > floor where the C++ keeps >= conf, so the floor is the largest Float below 0.001.
    public static let conf: Float = 0.001
    public static let confFloor: Float = Float(0.001).nextDown
    public static let iou: CGFloat = 0.7
    public static let maxDet = 300
}

public enum AccuracyRunner {
    public struct Outcome {
        public let map: MapResult
        public let inferMs: StageStats
        public let labels: String
        public var line: String { String(format: "[accuracy] images=%d  mAP50=%.4f  mAP50-95=%.4f", map.images, map.map50, map.map5095) }
        public func document() -> BenchDocument.Accuracy {
            BenchDocument.Accuracy(
                protocol: .init(conf: AccuracyProtocolValues.conf, iou: Float(AccuracyProtocolValues.iou),
                                max_det: AccuracyProtocolValues.maxDet, multi_label: true),
                labels: labels, images: map.images, map50: map.map50, map5095: map.map5095,
                per_class: map.perClass.map { .init(class_id: $0.cls, n_gt: $0.nGt, n_pred: $0.nPred, ap50: $0.ap50, ap5095: $0.ap5095) },
                timings: ["infer_ms": inferMs])
        }
    }

    /// "auto" = the ultralytics images -> labels rule, anything else is a labels directory.
    public static func labelsDir(_ spec: String) -> String? { spec == "auto" || spec.isEmpty ? nil : spec }

    /// Second pass over `images` at the val protocol; the boxes are scored with the txt rounding so
    /// the number equals scoring a --save-txt dump. `dump` receives every image's detections.
    public static func run(_ det: Detector, images: [URL], labels spec: String, resize: Int = 0,
                           dump: ((URL, [Detection]) -> Void)? = nil,
                           progress: ((Int, Int) -> Void)? = nil) -> Outcome {
        let ldir = labelsDir(spec)
        var evals: [ImageEval] = []
        evals.reserveCapacity(images.count)
        var infer: [Double] = []
        for (i, src) in images.enumerated() {
            autoreleasepool {
                guard var cg = loadCGImage(src) else { return }
                if resize > 0 { cg = resizeLong(cg, resize) }
                guard let res = try? det.detect(cg, conf: AccuracyProtocolValues.confFloor, iou: AccuracyProtocolValues.iou,
                                                maxDet: AccuracyProtocolValues.maxDet) else { return }
                infer.append(res.inferMs)
                dump?(src, res.detections)
                let gts = MapEvaluator.loadLabels(MapEvaluator.labelPath(forImage: src.path, labelsDir: ldir),
                                                  imageWidth: cg.width, imageHeight: cg.height)
                evals.append(ImageEval(preds: res.detections.map(TxtDump.predBox), gts: gts))
            }
            progress?(i + 1, images.count)
        }
        return Outcome(map: MapEvaluator.evaluate(evals), inferMs: StageStats(infer), labels: ldir ?? "auto")
    }
}
